In [2]:
print('hello')

hello


In [3]:
print('new')

new


Nota común

In [4]:
2+1

3

In [5]:
2-1

1

In [6]:
2*2

4

In [7]:
3/2

1.5

Modulo or "Mod" Operator

In [8]:
7/4

1.75

In [9]:
7%4

3

"Mod" me trae lo que resta luego de una división. 7 dividido 4, da 1 y un resto de 3, que dividido entre 4 me daría el .75

In [10]:
50%5

0

Puede servir para saber si un número es par o impar (sobre todo si ese valor esta escondido detras de una variabl), si da "0" al hace %2 entonces es par

Para elevar a una potencia se utiliza doble asterisco y la potencia **3 es al cubo

In [11]:
2**3

8

In [12]:
2 + 10 * 10 + 3

105

 Acá hace primero la multiplicación y luego la sumatoria, 10 * 10 = 100 -> 100 + 2 + 3 = 105

In [13]:
(2+10)*(10+3)

156

In [14]:
((2+3)**2)*(6/3)

50.0

In [16]:
(2+3)**2

25

In [17]:
((2+3)**2)*(2-(6/3))

0.0

In [18]:
((2+3)**2)*(2-(6/3)))

SyntaxError: unmatched ')' (Temp/ipykernel_4788/202838665.py, line 1)

In [20]:
((2+3)**2)*(9-(15/3))

100.0

In [21]:
0.1+0.2-0.3

5.551115123125783e-17

In [22]:
0.1 + 0.2 - 0.3

5.551115123125783e-17

In [23]:
0.1+0.2

0.30000000000000004

In [24]:
6*6

36

In [25]:
1/2

0.5

In [26]:
my_dog = 2

In [34]:
my_dog = 'Hermes'

In [35]:
my_dog

'Hermes'

In [36]:
a=5

In [37]:
a + a

10

In [47]:
a = a + a

In [42]:
a

20

In [43]:
a

20

In [45]:
a

40

In [48]:
type(a)

int

In [49]:
a 

160

In [50]:
a = 30.1

In [51]:
type(a)

float

In [52]:
my_income = 100

In [53]:
tax_rate = 0.1
my_taxes = my_income * tax_rate

In [55]:
my_taxes

10.0

In [56]:
'hello'
"world"

'world'

In [57]:
'this is also a string'

'this is also a string'

In [58]:
" I'm going on a run "

" I'm going on a run "

In [59]:
print ("hello")

hello


In [60]:
"hello world one"
"hello world two"

'hello world two'

In [61]:
print ("hello world one")
print ("hello world two")

hello world one
hello world two


In [62]:
print ("hello \n world")

hello 
 world


In [63]:
print ("hello \t world")

hello 	 world


In [64]:
print ("hello \nworld")

hello 
world


In [65]:
len ('hello')

5

In [66]:
my_dog

'Hermes'

In [67]:
len (my_dog)

6

In [104]:
my_name = 'Zalogon'

In [105]:
print (my_name)

Zalogon


In [106]:
len (my_name)

7

In [71]:
print ("Hello World")

Hello World


In [72]:
mystring = "Hello World"

In [73]:
mystring

'Hello World'

In [74]:
mystring[0]

'H'

In [75]:
mystring[8]

'r'

In [76]:
mystring[-2]

'l'

In [77]:
mystring ='abcdefghijk'

In [78]:
mystring[2:]

'cdefghijk'

In [79]:
mystring[:2]

'ab'

In [80]:
mystring[:3]

'abc'

In [81]:
 mystring[3:6]

'def'

In [83]:
mystring[2:9:2]

'cegi'

In [84]:
mystring[::2]

'acegik'

In [85]:
mystring[-10]

'b'

In [86]:
mystring[-10:10:-2]

''

In [87]:
mystring[-10:10]

'bcdefghij'

In [88]:
mystring[-10:10:2]

'bdfhj'

In [89]:
mystring[-10:10:-1]

''

In [90]:
mystring[1:-1:-2]

''

In [91]:
mystring[1:10:-2]

''

In [92]:
mystring[:-1:-2]

''

In [93]:
mystring[::-2]


'kigeca'

In [94]:
mystring = "Hello World"
mystring[8]

'r'

In [95]:
'tinker'[1:4]

'ink'

In [107]:
my_name

'Zalogon'

In [109]:
print(my_name[4:])

gon


In [114]:
print(my_name[:4])

Zalo


**IMMUTABILITY**

In [115]:
name = "Sam"

In [116]:
name[0] = "P"

TypeError: 'str' object does not support item assignment

In [117]:
# name[0] = "P"

In [118]:
name[1:]

'am'

In [119]:
last_letters = name[1:]

In [120]:
last_letters

'am'

In [121]:
'P' + last_letters

'Pam'

In [122]:
x = 'Hello World'

In [123]:
x + ' it is beautiful outside!'

'Hello World it is beautiful outside!'

In [124]:
x = x + ' it is beautiful outside!'

In [125]:
x

'Hello World it is beautiful outside!'

In [126]:
name

'Sam'

In [127]:
name - last_letters

TypeError: unsupported operand type(s) for -: 'str' and 'str'

In [128]:
b = name - last_letters

TypeError: unsupported operand type(s) for -: 'str' and 'str'

In [129]:
letter = 'z'

In [130]:
letter * 10

'zzzzzzzzzz'

In [131]:
2 + 3

5

In [132]:
'2' + '3'

'23'

In [133]:
newname = 'P'+last_letters

In [134]:
newname

'Pam'

In [135]:
x = 'Hello World'

In [136]:
x.upper ()

'HELLO WORLD'

In [137]:
x

'Hello World'

In [138]:
x.upper

<function str.upper()>

In [139]:
x.lower()

'hello world'

In [140]:
x.split()

['Hello', 'World']

In [141]:
x = 'This is a string'

In [142]:
x.split('i')

['Th', 's ', 's a str', 'ng']

In [144]:
x.swapcase()

'tHIS IS A STRING'

In [145]:
print('Hello ' + my_name)

Hello Zalogon


In [146]:
print('This is a string {}'.format('INSERTED'))

This is a string INSERTED


In [147]:
 print('The {} is {} and {}.'.format('fox', 'brown', 'quick'))

The fox is brown and quick.


In [149]:
 print('The {2} {1} {0}.'.format('fox', 'brown', 'quick'))

The quick brown fox.


In [150]:
 print('The {0} {0} {0}.'.format('fox', 'brown', 'quick'))

The fox fox fox.


In [151]:
 print('The {2} {1} {0}.'.format(my_name, 'brown', 'quick'))

The quick brown Zalogon.


In [152]:
 print('The {q} {b} {f}.'.format(f='fox',b='brown',q='quick'))

The quick brown fox.


**Float formating follow "{value:width.precision *f* }"** 

In [154]:
result = 100/777

In [155]:
result

0.1287001287001287

In [157]:
print("The result was {r}".format(r=result))

The result was 0.1287001287001287


In [160]:
print("The result was {r:1.3}".format(r=result))

The result was 0.129


In [161]:
print("The result was {r:1.3f}".format(r=result))

The result was 0.129


In [165]:
print("The result was {r:10.3f}".format(r=result))

The result was      0.129


In [166]:
print("The result was {r:10.5f}".format(r=result))

The result was    0.12870


In [167]:
result = 104.12345

In [170]:
print("The result was {r:15.3f}".format(r=result))

The result was         104.123


In [171]:
print("The result was {r:1.5f}".format(r=result))

The result was 104.12345


**f-stings method** - *formated strings literals*

Se agrega una **f** antes del string y luego se anota entre llaves directo lo que querés que te traiga

In [172]:
name = "Jose"

In [173]:
print(f'Hello,his name is {name}')

Hello,his name is Jose


In [174]:
 age = 3

In [176]:
print(f'{name} is {age} years old.')

Jose is 3 years old.


In [179]:
'{} rules!'.format('Python')

'Python rules!'


**LISTS**


In [180]:
my_list = [1,2,3]

In [181]:
my_list = ['STRING', 100, 23.2]

In [182]:
len(my_list)

3

In [183]:
my_list = ['one', 'two', 'three']

In [184]:
my_list[0]

'one'

In [185]:
my_list[1:]

['two', 'three']

In [186]:
my_list

['one', 'two', 'three']

In [187]:
another_list=['four', 'five']

In [188]:
my_list + another_list

['one', 'two', 'three', 'four', 'five']

In [189]:
new_list = my_list + another_list

In [190]:
new_list[0] = 'ONE ALL CAPS'

In [191]:
my_list

['one', 'two', 'three']

In [192]:
new_list

['ONE ALL CAPS', 'two', 'three', 'four', 'five']

In [193]:
my_list

['one', 'two', 'three']

*Se pueden agregar elementos al final de una lista usando* **.append**


In [194]:
new_list.append('six')

In [195]:
new_list

['ONE ALL CAPS', 'two', 'three', 'four', 'five', 'six']

In [196]:
new_list.append('seven')

In [197]:
new_list

['ONE ALL CAPS', 'two', 'three', 'four', 'five', 'six', 'seven']

*Se pueden quitar elementos al final de una lista usando* **.pop**



In [198]:
new_list.pop()

'seven'

In [199]:
new_list

['ONE ALL CAPS', 'two', 'three', 'four', 'five', 'six']

In [200]:
popped_item = new_list.pop()

In [201]:
popped_item

'six'

In [202]:
new_list

['ONE ALL CAPS', 'two', 'three', 'four', 'five']

In [203]:
new_list.pop(0)

'ONE ALL CAPS'

In [204]:
new_list

['two', 'three', 'four', 'five']

In [205]:
new_list = ['a', 'e', 'x', 'b', 'c']

In [206]:
num_list = [4,1,8,3]

In [207]:
new_list.sort()

In [208]:
new_list

['a', 'b', 'c', 'e', 'x']

In [209]:
my_sorted_list = new_list.sort()

In [210]:
my_sorted_list

In [211]:
type(my_sorted_list)

NoneType

In [212]:
None

"None" es un default value para funciones que no traen ningun resultado, como .sort() ordena la lista pero no trae ningun valor, por eso "my_sorted_list = new_list.sort()" no trae nada y Python la reconoce como un NoneType

In [213]:
num_list

[4, 1, 8, 3]

In [214]:
num_list.sort()

In [215]:
num_list

[1, 3, 4, 8]

In [216]:
num_list.reverse()

In [217]:
num_list

[8, 4, 3, 1]

In [218]:
num_list[1:]

[4, 3, 1]

In [219]:
num_list[2]

3

In [1]:
lst=[0,1,2]

In [2]:
lst.pop()

2

In [5]:
my_dict = {'key1' : 'value1' ,  'key2' : 'value2'}

In [6]:
my_dict

{'key1': 'value1', 'key2': 'value2'}

In [7]:
my_dict['key1']

'value1'

In [8]:
prices_lookup = {'apple' : 2.99, 'oranges':1.99, 'milk':5.80}

In [9]:
 prices_lookup['apple']

2.99

In [3]:
d = {'k1': 123, 'k2':[0,1,2], 'k3':{'insideKey':100}}

In [4]:
d['k2']

[0, 1, 2]

In [5]:
d['k3']

{'insideKey': 100}

In [6]:
d['k3']['insideKey']

100

In [7]:
d['k2']

[0, 1, 2]

In [8]:
d['k1']

123

In [10]:
d['k2'][1]

1

In [18]:
d = {'key1':['a', 'b', 'c']}

In [21]:
d

{'key1': ['a', 'b', 'c']}

In [22]:
d['key1'][2]

'c'

In [26]:
d['key1'][2].upper()

'C'

In [28]:
d = {'k1': 100, 'k2':200}

In [29]:
d['k3']=300

In [30]:
d

{'k1': 100, 'k2': 200, 'k3': 300}

In [33]:
d['k1']='NEW VALUE'

In [32]:
d

{'k1': 666, 'k2': 200, 'k3': 300}

In [34]:
d = {'k1': 100, 'k2': 200, 'k3': 300}

In [35]:
d.keys()

dict_keys(['k1', 'k2', 'k3'])

In [36]:
d.values()

dict_values([100, 200, 300])

In [37]:
d.items()

dict_items([('k1', 100), ('k2', 200), ('k3', 300)])

**KEYS son siempre Strings**

**TUPLES** son inmutables, usan ()

In [38]:
t = (1, 2, 3)

In [39]:
mylist = [1,2,3]

In [40]:
type(t)

tuple

In [41]:
type(mylist)

list

In [42]:
len(t)

3

In [43]:
t

(1, 2, 3)

In [45]:
t = ('one', 2)

In [46]:
t[0]

'one'

In [47]:
t[-1]

2

In [48]:
t = ('a', 'a', 'b')

In [49]:
t.count('a')

2

In [50]:
t.index('a')

0

Con .index si hay repetidos solo me trae el primero

In [51]:
t.index('b')

2

In [52]:
t

('a', 'a', 'b')

In [53]:
mylist

[1, 2, 3]

In [56]:
mylist[0]='NEW'

In [57]:
mylist

['NEW', 2, 3]

In [58]:
t[0]= 'NEW'

TypeError: 'tuple' object does not support item assignment

In [59]:
t=(1,2,[1,2])

In [60]:
t

(1, 2, [1, 2])


   **SETS** unordered collection of **unique** elements
   

In [1]:
 myset = set()

In [2]:
myset

set()

In [3]:
myset.add(1)

In [4]:
myset

{1}

In [5]:
myset.add(2)

In [6]:
myset

{1, 2}

In [7]:
myset.add(2)

In [8]:
myset

{1, 2}

Puede servir para meter una lista dentro de un SET y asi obtener solo los valores unicos, sin los repetidos

In [9]:
mylist= (1,1,1,1,2,2,2,2,3,3,3,3,)

In [10]:
set(mylist)

{1, 2, 3}

In [11]:
set

set

In [14]:
mylist= (3,3,'asda', 'asda', 'aaa', 3.4,['b','a', 'c', 'c'],[3,'asda', 3.4, 4.5, 'b'])

In [15]:
set(mylist)

TypeError: unhashable type: 'list'

In [16]:
mylist

(3,
 3,
 'asda',
 'asda',
 'aaa',
 3.4,
 ['b', 'a', 'c', 'c'],
 [3, 'asda', 3.4, 4.5, 'b'])

In [17]:
set(mylist)

TypeError: unhashable type: 'list'

In [18]:
set('Mississippi')

{'M', 'i', 'p', 's'}


   **Booleans** operador que permite trasmitir afirmaciones **True** o **False** (Siempre con **T** y **F** mayúsculas)
    
    (Sirven para el control de flujo y lógica)

In [19]:
True

True

In [20]:
true

NameError: name 'true' is not defined

In [21]:
False

False

In [22]:
type(False)

bool

In [23]:
1 > 2

False

In [24]:
1 < 1

False

In [25]:
1 == 1

True

In [26]:
1 =< 1

SyntaxError: cannot assign to literal (Temp/ipykernel_8540/686455306.py, line 1)

In [27]:
b = None

In [28]:
b


   **I/O** Input and Output with basic .txt files
    

In [2]:
test.txt

NameError: name 'test' is not defined

 SOLO en Jupyter Notebook se puede usar
 %%writefile myfile.txt
 Hello this is a text file
 this is the second line
 this is the third line

In [3]:
%%writefile myfile.txt 
Hello this is a text file 
this is the second line 
this is the third line

Writing myfile.txt


In [4]:
myfile = open('myfile.txt')

In [5]:
myfile = open('whoops_wrong.txt')

FileNotFoundError: [Errno 2] No such file or directory: 'whoops_wrong.txt'

In [7]:
test = open ('test.txt')

In [8]:
pwd

'C:\\Users\\Gin\\Desktop\\PYTHON\\Python - Zero-2-Hero'

 Puedo usar **pwd** (*print work directory*) para saber donde está ubicada esta notebook

In [9]:
test.txt

AttributeError: '_io.TextIOWrapper' object has no attribute 'txt'

In [10]:
myfile.txt

AttributeError: '_io.TextIOWrapper' object has no attribute 'txt'

In [11]:
test = open ('test.txt')

In [12]:
myfile = open('myfile.txt')

In [13]:
myfile.read()

'Hello this is a text file \nthis is the second line \nthis is the third line\n'

Si uso **.read()** me trae todo como una sola string, usando **\n** para indicar una *nueva linea*

In [14]:
myfile.read()

''

Esto pasa porque el "Cursor" fue hasta el final del archivo
Para resetear eso puedo usar **seek(0)**

In [15]:
myfile.seek(0)

0

In [16]:
myfile.read()

'Hello this is a text file \nthis is the second line \nthis is the third line\n'

In [17]:
myfile.seek(8)

8

In [18]:
myfile.read()

'is is a text file \nthis is the second line \nthis is the third line\n'

In [19]:
myfile.seek(0)

0

In [20]:
contents = myfile.read()

In [21]:
contents

'Hello this is a text file \nthis is the second line \nthis is the third line\n'

In [22]:
myfile.seek(0)

0

In [23]:
myfile.read(4)

'Hell'

In [24]:
myfile.seek(0)

0

In [25]:
myfile.read(:4)

SyntaxError: invalid syntax (Temp/ipykernel_7464/3709802497.py, line 1)

In [26]:
myfile.seek(0)

0

In [27]:
myfile.read([:4])

SyntaxError: invalid syntax (Temp/ipykernel_7464/2970538716.py, line 1)

In [28]:
myfile.read()

'Hello this is a text file \nthis is the second line \nthis is the third line\n'

In [29]:
myfile.seek(0)

0

In [30]:
myfile.readlines()

['Hello this is a text file \n',
 'this is the second line \n',
 'this is the third line\n']

In [31]:
myfile = open('C:\\Users\\Gin\\Desktop\\PYTHON\\Python - Zero-2-Hero\\test.txt')


**Best Practice** Hay que cerrar el archivo que abrimos cuando termino de usarlo. Se usa **.close()**

In [32]:
myfile.close()

In [33]:
test.close()

In [34]:
myfile = open (myfile.txt)

AttributeError: '_io.TextIOWrapper' object has no attribute 'txt'

In [35]:
myfile = open('myfile.txt')

In [36]:
myfile.close()


Puedo abrirlo como uno nuevo usando **with** antes de open('myfile.txt') **as** my_new_file:

    <- Esto genera un "indented" block of code que va dentro my_new_file en lugar de myfile

In [38]:
with open ('myfile.txt') as my_new_file:
    contents = my_new_file.read()

In [39]:
contents

'Hello this is a text file \nthis is the second line \nthis is the third line\n'


Para escribir puedo usar **with open('myfile.text',)** al agregar la coma voy con el cursor al espacio entre **(** y **'** , luedo de **open** y hago **shift + tab** para opciones y revisar con que mode quiero abrirlo indicandolo luego de la coma:

   **with open('myfile.text',mode='r'** para read

In [42]:
with open('myfile.txt',mode='r') as myfile:
    contents = myfile.read()

In [43]:
with open('myfile.txt',mode='w') as myfile:
    contents = myfile.read()

UnsupportedOperation: not readable

Dependiendo del modo que elijo tal vez no puedo hacer lo que quiero, ej: si lo abro para escribir no lo puedo usar para leer

...because of permissions

In [44]:
with open('myfile.txt',mode='a') as myfile:
    contents = myfile.read()

UnsupportedOperation: not readable

**Reading, Writing and Appending Modes**

**· mode='r'**       open for reading only (default)

**· mode='w'**       open for writing only (will overwrite files or create new!)

**· mode='a'**       open for appending only (will add on to files)

**· mode='r+'**      is reading and writing

**· mode='w+'**      is writing and read (Overwrites existing files or creates a new file!)



Tambien:
'x'       create a new file and open it for writing
'b'       binary mode
't'       text mode (default)
'+'       open a disk file for updating (reading and writing)
'U'       universal newline mode (deprecated)


In [53]:
%%writefile my_new_file.txt
ONE ON FIRST
TWO ON SECOND
THREE ON THIRD

Overwriting my_new_file.txt


In [54]:
with open('my_new_file.txt',mode='r') as f:
    print(f.read())

ONE ON FIRST
TWO ON SECOND
THREE ON THIRD



In [55]:
with open('my_new_file.txt',mode='a') as f:
    f.write('FOUR ON FOURTH')

In [56]:
with open('my_new_file.txt',mode='r') as f:
    print(f.read())

ONE ON FIRST
TWO ON SECOND
THREE ON THIRD
FOUR ON FOURTH


In [58]:
with open('asdfghjkl.txt', mode='w') as f:
    f.write('I CREATED THIS FILE!')

In [61]:
with open('asdfghjkl.txt', mode='r') as f:
    print(f.read())

I CREATED THIS FILE!


# Comparison Operators 

In this lecture we will be learning about Comparison Operators in Python. These operators will allow us to compare variables and output a Boolean value (True or False). 

If you have any sort of background in Math, these operators should be very straight forward.

First we'll present a table of the comparison operators and then work through some examples:

<h2> Table of Comparison Operators </h2><p>  In the table below, a=3 and b=4.</p>

<table class="table table-bordered">
<tr>
<th style="width:10%">Operator</th><th style="width:45%">Description</th><th>Example</th>
</tr>
<tr>
<td>==</td>
<td>If the values of two operands are equal, then the condition becomes true.</td>
<td> (a == b) is not true.</td>
</tr>
<tr>
<td>!=</td>
<td>If values of two operands are not equal, then condition becomes true.</td>
<td>(a != b) is true</td>
</tr>
<tr>
<td>&gt;</td>
<td>If the value of left operand is greater than the value of right operand, then condition becomes true.</td>
<td> (a &gt; b) is not true.</td>
</tr>
<tr>
<td>&lt;</td>
<td>If the value of left operand is less than the value of right operand, then condition becomes true.</td>
<td> (a &lt; b) is true.</td>
</tr>
<tr>
<td>&gt;=</td>
<td>If the value of left operand is greater than or equal to the value of right operand, then condition becomes true.</td>
<td> (a &gt;= b) is not true. </td>
</tr>
<tr>
<td>&lt;=</td>
<td>If the value of left operand is less than or equal to the value of right operand, then condition becomes true.</td>
<td> (a &lt;= b) is true. </td>
</tr>
</table>

In [1]:
2 == 2

True

In [2]:
1 == 2

False

In [3]:
'hello' == 'bye'

False

In [4]:
'Bye' == 'bye'

False

In [5]:
2.0 == 2

True

In [6]:
2 == '2'

False

In [7]:
3 != 3

False

In [8]:
4 != 5

True

In [9]:
2 < 1

False

In [10]:
2 > 1

True

In [11]:
2 <= 2

True

In [12]:
3 <= 2

False

In [13]:
1 <= 2

True

In [15]:
4 >= 3

True

**Chaining comaparison operators**

**and**

**or**

**not** brings the opposite bool of the comparison (if it 1 == 1 is True, "not 1==1" brings False)

In [16]:
1 < 2 < 3

True

In [17]:
1 < 2 > 3

False

In [18]:
1 < 2 and 3 >= 3

True

In [19]:
2 > 1 or 4 == 5

True

In [20]:
'hello' == 'hello' or 1 == 1

True

In [21]:
'hello' != 'hello' or 1 == 2

False

In [22]:
1 == 1

True

In [23]:
not (1==1)

False

In [24]:
not 1==1

False

In [25]:
123 == 5

False

In [26]:
not 123 == 5

True

# if, elif, else Statements

<code>if</code> Statements in Python allows us to tell the computer to perform alternative actions based on a certain set of results.

Verbally, we can imagine we are telling the computer:

"Hey if this case happens, perform some action"

We can then expand the idea further with <code>elif</code> and <code>else</code> statements, which allow us to tell the computer:

"Hey if this case happens, perform some action. Else, if another case happens, perform some other action. Else, if *none* of the above cases happened, perform this action."

Let's go ahead and look at the syntax format for <code>if</code> statements to get a better idea of this:

    if case1:
        perform action1
    elif case2:
        perform action2
    else: 
        perform action3

In [28]:
if True:
    print ('ITS TRUE')

ITS TRUE


In [29]:
if 3<4:
    print ('ITS TRUE')

ITS TRUE


In [30]:
hungry = True
if hungry:
    print('Feed me!')

Feed me!


In [32]:
hungry = False
if hungry:
    print('Feed me!')
else:
    print('Im not hungry')

Im not hungry


In [33]:
loc = 'Bank'

if loc == 'Auto Shop':
    print ('Cars are cool!')
else:
    print ('I dont know munch')

I dont know munch


In [35]:
loc = 'Game'

if loc == 'Auto Shop':
    print ('Cars are cool!')
elif loc == 'bank':
    print ('Money is cool!')
elif loc =='Store':
    print ('Welcome to the store!')
else:
    print ('I dont know much')

I dont know munch


In [37]:
name = 'Ramirez'

if name == 'Frankie':
    print ('Hello Frankie')
elif name == 'Sammy':
    print ('Good day Sammy')
else:
    print(('Wellmet ')+name)

Wellmet Ramirez


# for Loops

A <code>for</code> loop acts as an iterator in Python; it goes through items that are in a *sequence* or any other iterable item. Objects that we've learned about that we can iterate over include strings, lists, tuples, and even built-in iterables for dictionaries, such as keys or values.

We've already seen the <code>for</code> statement a little bit in past lectures but now let's formalize our understanding.

Here's the general format for a <code>for</code> loop in Python:

    for item in object:
        statements to do stuff
    

The variable name used for the item is completely up to the coder, so use your best judgment for choosing a name that makes sense and you will be able to understand when revisiting your code. This item name can then be referenced inside your loop, for example if you wanted to use <code>if</code> statements to perform checks.

Let's go ahead and work through several example of <code>for</code> loops using a variety of data object types. We'll start simple and build more complexity later on.

## Example 1
Iterating through a list

In [40]:
my_pets = ['Coco', 'Leeloo', 'Hermes']
for pet_name in my_pets:
    print (pet_name)

Coco
Leeloo
Hermes


In [49]:
mylist = [1,2,3,4,5,6,7,8,9,10]

In [43]:
for number in mylist:
    print(number)
    
    # ponerle "number" puede ser cualquier variable name, pero va a representar cada nùmero dentro de esa lista.

1
2
3
4
5
6
7
8
9
10


In [44]:
for number in mylist:
    print('Hello')

Hello
Hello
Hello
Hello
Hello
Hello
Hello
Hello
Hello
Hello


In [46]:
for number in mylist:
    # Check for even
    if number % 2 == 0:
        print (number)
    else:
        print (f'Odd Number: {number}')

Odd Number: 1
2
Odd Number: 3
4
Odd Number: 5
6
Odd Number: 7
8
Odd Number: 9
10


In [51]:
list_sum = 0

for number in mylist:
    list_sum = list_sum + number

print(list_sum)

55


In [52]:
list_sum = 0

for number in mylist:
    list_sum = list_sum + number

    print(list_sum)

1
3
6
10
15
21
28
36
45
55


In [54]:
mystring = 'Praise the Omnissiah!'
for letter in mystring:
    print (letter)

P
r
a
i
s
e
 
t
h
e
 
O
m
n
i
s
s
i
a
h
!


In [56]:

for asdf in 'Bubulele':
    print (asdf)

B
u
b
u
l
e
l
e


In [57]:
#algunos usan el underscore/guiòn bajo _ para indicar en la función un iterable que no piensa usar luego, y no se gastan en ponerle un nombre acorde
for _ in 'Bubulele':
    print ('Cool!')

Cool!
Cool!
Cool!
Cool!
Cool!
Cool!
Cool!
Cool!


In [59]:
tup = (1,2,3)

for item in tup:
    print (item)

1
2
3


**Tuples** have a special quality when it comes to for loops. If you are iterating through a sequence that contains tuples, *the item can actually be the tuple itself*, this is an example of **tuple unpacking**. During the for loop we will be unpacking the tuple inside of a sequence and we can access the individual items inside that tuple!

In [60]:
mylist = [(1,2), (3,4), (5,6), (7,8)]

In [61]:
len (mylist)

4

In [62]:
for item in mylist:
    print(item)

(1, 2)
(3, 4)
(5, 6)
(7, 8)


In [63]:
for (a,b) in mylist:
    print(a)
    print(b)

1
2
3
4
5
6
7
8


Duplicas las estructura de los items dentro de la secuencia para el unpacking

In [64]:
for a,b in mylist:
    print(a)

1
3
5
7


De hecho no necesita los ()

In [71]:
x = 'sun'
y = 777
z = 4.5
mylist = [(1,2,3),(x,y,z),('alpha','beta', 'gamma')]

In [72]:
for a,b,c in mylist:
    print(a)
    print(b)
    print(c)

1
2
3
sun
777
4.5
alpha
beta
gamma


In [74]:
d = {'k1':1, 'k2':2, 'k3':3}

for item in d:
    print (item)

k1
k2
k3


In [75]:
d = {'k1':1, 'k2':2, 'k1':3}

for item in d:
    print (item)

k1
k2


In [76]:
mylist = [1,2,32,13,2,1,4,1,2,3,3]
for number in mylist:
    print(number)

1
2
32
13
2
1
4
1
2
3
3


In [77]:
d = {'k1':1, 'k2':2, 'k3':3}

for item in d.items():
    print (item)

('k1', 1)
('k2', 2)
('k3', 3)


In [78]:
d = {'k1':1, 'k2':2, 'k3':3}

for key,value in d.items():
    print (value)

1
2
3


As dictionaries and unordered, when a iterate there is no guarantee that it will return the values in the same order they where introduced to the dictionary.

In [1]:
#

# while Loops

The <code>while</code> statement in Python is one of most general ways to perform iteration. A <code>while</code> statement will repeatedly execute a single statement or group of statements as long as the condition is true. The reason it is called a 'loop' is because the code statements are looped through over and over again until the condition is no longer met.

The general format of a while loop is:

    while test:
        code statements
    else:
        final code statements

Let’s look at a few simple <code>while</code> loops in action. 

# break, continue, pass

We can use <code>break</code>, <code>continue</code>, and <code>pass</code> statements in our loops to add additional functionality for various cases. The three statements are defined by:

    break: Breaks out of the current closest enclosing loop.
    continue: Goes to the top of the closest enclosing loop.
    pass: Does nothing at all.
    
    
Thinking about <code>break</code> and <code>continue</code> statements, the general format of the <code>while</code> loop looks like this:

    while test: 
        code statement
        if test: 
            break
        if test: 
            continue 
    else:

<code>break</code> and <code>continue</code> statements can appear anywhere inside the loop’s body, but we will usually put them further nested in conjunction with an <code>if</code> statement to perform an action based on some condition.

In [2]:
x = 0

while x < 5:
    print(f'The current value of x is {x}')
    x = x + 1 #can also write "x += 1"

The current value of x is 0
The current value of x is 1
The current value of x is 2
The current value of x is 3
The current value of x is 4


In [4]:
x = 0

while x < 5:
    print(f'The current value of x is {x}')
    x += 1

else:
    print ('X is not less than 5')

The current value of x is 0
The current value of x is 1
The current value of x is 2
The current value of x is 3
The current value of x is 4
X is not less than 5


In [6]:
x = (1,2,3)

for item in x:
    # comment
    pass

print ('end of my script')

end of my script


In [7]:
mystring = 'Sammy'

In [8]:
for letter in mystring:
    print (letter)

S
a
m
m
y


In [9]:
for letter in mystring:
    if letter == 'a':
        continue
    print (letter)

S
m
m
y


In [10]:
for letter in mystring:
    if letter == 'a':
        break
    print (letter)

S


In [11]:
for letter in mystring:
    if letter == 'a':
        pass
    print (letter)

S
a
m
m
y


In [14]:
x = 0

while x < 5:
    print (x)
    x += 1

0
1
2
3
4


In [16]:
x = 0

while x < 9:
    if x == 2:
        break
    print(x)
    x += 1

0
1


In [17]:
#


# Useful Operators

There are a few built-in functions and "operators" in Python that don't fit well into any category, so we will go over them in this lecture, let's begin!

## range

The range function allows you to quickly *generate* a list of integers, this comes in handy a lot, so take note of how to use it! There are 3 parameters you can pass, a start, a stop, and a step size. Let's see some examples:

In [18]:
mylist = [1,2,3]

In [20]:
for num in range(3,10):
    print (num)

3
4
5
6
7
8
9


In [21]:
for num in range(3,10, 2):
    print (num)

3
5
7
9


In [22]:
for num in range(10):
    print (num)

0
1
2
3
4
5
6
7
8
9


In [23]:
range (10)

range(0, 10)

In [24]:
list(range(0,11,2))

[0, 2, 4, 6, 8, 10]

## enumerate

enumerate is a very useful function to use with for loops.

In [25]:
for i,letter in enumerate('abcde'):
    print("At index {} the letter is {}".format(i,letter))

At index 0 the letter is a
At index 1 the letter is b
At index 2 the letter is c
At index 3 the letter is d
At index 4 the letter is e


In [26]:
index_count = 0

for letter in 'abcde':
    print ('At index {} the letter is {}'.format(index_count,letter))
    index_count += 1

At index 0 the letter is a
At index 1 the letter is b
At index 2 the letter is c
At index 3 the letter is d
At index 4 the letter is e


In [27]:
index_count = 0
word = 'abcde'
for letter in word:
    print (word[index_count])
    index_count+=1

a
b
c
d
e


In [28]:
word = 'abcde'
for item in enumerate (word):
    print(item)

(0, 'a')
(1, 'b')
(2, 'c')
(3, 'd')
(4, 'e')


In [29]:
word = 'abcde'
for index,letter in enumerate (word):
    print(index)
    print(letter)
    print ('\n')

0
a


1
b


2
c


3
d


4
e




## zip

Notice the format enumerate actually returns, let's take a look by transforming it to a list()

In [30]:
list(enumerate('abcde'))

[(0, 'a'), (1, 'b'), (2, 'c'), (3, 'd'), (4, 'e')]

It was a list of tuples, meaning we could use tuple unpacking during our for loop. This data structure is actually very common in Python , especially when working with outside libraries. You can use the zip() function to quickly create a list of tuples by "zipping" up together two lists.

In [22]:
mylist1 = [1,2,3,4,5]

In [23]:
mylist2= ['a', 'b', 'c']

In [24]:
zip(mylist1, mylist2)

In [25]:
for item in zip(mylist1, mylist2):
    print(item)

(1, 'a')
(2, 'b')
(3, 'c')


In [26]:
mylist3 = ['blue', 'red', 'white']


In [27]:
for item in zip(mylist1, mylist2, mylist3):
    print(item)

(1, 'a', 'blue')
(2, 'b', 'red')
(3, 'c', 'white')


**It will unify list up to the shortest one**

In [28]:
list(zip(mylist1, mylist2, mylist3))

[(1, 'a', 'blue'), (2, 'b', 'red'), (3, 'c', 'white')]

In [33]:
for num, let, col in zip(mylist1, mylist2, mylist3):
    print ("I have {} shirt/s with a {} letter {} on it's fornt".format(num, col, let))

I have 1 shirt/s with a blue letter a on it's fornt
I have 2 shirt/s with a red letter b on it's fornt
I have 3 shirt/s with a white letter c on it's fornt



## in operator

Can use **in** to quickly check if an object is in a list

It returns a True or False result

In [34]:
'x' in [1,2,3]

False

In [36]:
'x' in ('x','y','z') #works with tuples

True

In [37]:
'x' in {'a' : 1, 'b' : 'blue' , 'c' : mylist1}

False

In [44]:
'x' in {'a' : 1, 'x' : 'blue' , 'c' : mylist1} #works with Keys in Dictionaries, not with the Value asociated with that Key unless you use .values()

True

In [45]:
'x' in {'a' : 1, 'b' : 'x' , 'c' : mylist1}

False

In [40]:
x = 'whatever'

In [41]:
'x' in {'a' : 1, 'b' : x , 'c' : mylist1}

False

In [42]:
'a' in 'Omnissiah'

True

In [46]:
d = {'a' : 1, 'b' : 'x' , 'c' : mylist1}
'x' in d.values()

True

## not in

We can combine **in** with a **not** operator, to check if some object or variable is not present in a list.


In [48]:
'x' not in [1,2,3]

True

In [49]:
'x' not in ('x','y','z')

False

## min and max

Quickly check the minimum or maximum of a list with these functions.

In [58]:
mylist = [10,200,30,4,100]

In [59]:
min(mylist)

4

In [60]:
max(mylist)

200

In [61]:
t = (1,2,3,4,5,9,0)

In [62]:
min(t)

0

In [63]:
max(t)

9

In [56]:
min (d)

'a'

In [64]:
max (d)

'c'

In [65]:
d2={'keyZ': '300', 'keyAloha': 43, 'keyFirst': 'a', 'keyMiddle': '1','keyLast' : 'z'}

In [66]:
min (d2)

'keyAloha'

In [67]:
max (d2)

'keyZ'

In [68]:
d3={'key1': '300', 'zkeyAloha': 43, 'keyFirst': 'a', 'keyMiddle': '1','keyLast' : 'z'}

In [69]:
min(d3)

'key1'

In [70]:
max(d3)

'zkeyAloha'

In [72]:
max d2.values()

SyntaxError: invalid syntax (Temp/ipykernel_10716/2258666371.py, line 1)

## random

Python comes with a built in random library. There are a lot of functions included in this random library, so we will only show you two useful functions for now.

In [75]:
from random import shuffle

# This shuffles the list "in-place" meaning it won't return
# anything, instead it will effect the list passed

shuffle(mylist1)

In [76]:
mylist1

[5, 2, 4, 3, 1]

In [77]:
shuffle(mylist1)
mylist1

[4, 3, 2, 1, 5]

In [78]:
from random import randint

# Return random integer in range [a, b], including both end points.

randint(0,100)

15

In [79]:
randint(0,100)

77

In [82]:
t = (1,2,3,4,5,6,7,8,9,10)
from random import shuffle
shuffle(t)

TypeError: 'tuple' object does not support item assignment

In [84]:
mylist_formtuple = [t]

In [85]:
from random import shuffle
shuffle(mylist_formtuple)

In [86]:
mylist_formtuple

[(1, 2, 3, 4, 5, 6, 7, 8, 9, 10)]

 Shuffle no funciona con Tuples

In [87]:
from random import randint
print ('1d20=',randint(1,20))

1d20= 9


In [94]:
from random import randint
print ('3d6=',(randint(1,6))+(randint(1,6))+(randint(1,6)))

3d6= 10


In [95]:
(randint(1,6))+(randint(1,6))

5

In [100]:
dicepool = 10
roll = []
roll.append((randint(1,10))*dicepool)
roll

[90]

In [101]:
dicepool = 7
roll = []
roll.append((randint(1,10))*dicepool)
roll

[35]

In [110]:
dicepool = 7
roll = []
roll.append((randint(1,10))*dicepool)
roll

[28]

In [103]:
myroll = roll

In [104]:
myroll

[49]

## input

In [105]:
input ('Box to type things: ')

Box to type things: Praise the Omnissiah!


'Praise the Omnissiah!'

Always returns a string, even it the input in the box was a number (it won't recognize it an integer)

In [111]:
dicepool = input ("What's your dice pool?")

What's your dice pool?5


In [112]:

roll = []
roll.append((randint(1,10))*dicepool)
roll

['5555']

In [113]:
dicepool = input ("What's your dice pool?")

What's your dice pool?3


In [116]:
type(dicepool)


str

In [117]:
float(dicepool)

3.0

In [118]:
int(dicepool)

3

In [119]:
roll = []
roll.append((randint(1,10))*(int (dicepool)))
roll

[18]

In [120]:
roll = []
roll.append((randint(1,10))*(int (dicepool)))
roll

[18]

In [121]:
roll = []
roll.append((randint(1,10))*(int (dicepool)))
roll

[15]

In [126]:
roll = []
roll.append((randint(1,10))*(int (dicepool)))
roll

[12]

In [127]:
# I can transform it into a integer in the same line

dicepool = int(input ("What's your dice pool?"))

What's your dice pool?7


In [128]:
type(dicepool)

int

In [129]:
dicepool

7

In [134]:
roll = []
roll.append((randint(1,10))*(dicepool))
roll

[63]

In [136]:
roll = []
dicepool*roll.append(randint(1,10))
roll

TypeError: unsupported operand type(s) for *: 'int' and 'NoneType'

In [143]:
roll = []
diceadded = 0


In [145]:
while diceadded <= dicepool:
    roll.append(randint(1,10)
    diceadded = diceadded+1
    print ('We have 'diceadded' dice added to the pool')
    
else:
    print (roll)

SyntaxError: invalid syntax (Temp/ipykernel_10716/850097591.py, line 3)

In [161]:
from random import randint
dicepool = int(input ("What's your dice pool?"))
roll = []
diceadded = 0
while diceadded <= dicepool:
    diceadded = diceadded+1
    roll.append(randint(1,10))
                
else:
    print (roll)

What's your dice pool?8
[7, 1, 4, 2, 10, 5, 2, 9, 8]


In [162]:
from random import randint
dicepool = int(input ("What's your dice pool?"))
roll = []
diceadded = 0
while diceadded <= dicepool:
    diceadded = diceadded+1
    roll.append(randint(1,10))
                
else:
    print (roll)

What's your dice pool?4
[6, 6, 6, 1, 8]


In [163]:
from random import randint
dicetype = int(input ("How many faces do your dice have? "))
dicepool = int(input ("What's your dice pool? "))
roll = []
diceadded = 0
while diceadded < dicepool:
    diceadded = diceadded+1
    roll.append(randint(1,(dicetype)))
                
else:
    print ('You rolled {}d{}'.format (dicepool,dicetype))
    print ('Your results are:{}'.format(roll))

How many faces do your dice have? 10
What's your dice pool? 6
You rolled 6d10
Your results are:[5, 6, 7, 9, 5, 8]


# List Comprehensions

In addition to sequence operations and list methods, Python includes a more advanced operation called a list comprehension.

List comprehensions allow us to build out lists using a different notation. You can think of it as essentially a one line <code>for</code> loop built inside of brackets. For a simple example:
## Example 1

In [1]:
mystring = 'hello'

In [2]:
mylist = []

for letter in mystring:
    mylist.append(letter)

In [3]:
mylist

['h', 'e', 'l', 'l', 'o']

In [6]:
mylist = []

In [7]:
mylist

[]

In [8]:
mylist = [letter for letter in mystring]


In [9]:
mylist

['h', 'e', 'l', 'l', 'o']

In [10]:
mylist = [x for x in 'word']

In [11]:
mylist

['w', 'o', 'r', 'd']

In [12]:
mylist = [blahblah for blahblah in 'iterable object']

In [13]:
mylist

['i', 't', 'e', 'r', 'a', 'b', 'l', 'e', ' ', 'o', 'b', 'j', 'e', 'c', 't']

In [14]:
mylist = [num for num in range (0,11)]

In [15]:
mylist

[0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]

One can use operations in that first element

In [16]:
mylist = [num**2 for num in range (0,11)]

In [17]:
mylist

[0, 1, 4, 9, 16, 25, 36, 49, 64, 81, 100]

In [18]:
mylist = ['a' for num in range (0,11)]

In [19]:
mylist

['a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a', 'a']

Can also use **if** statements

In [20]:
mylist = [x for x in range (0,11) if x%2==0]  

In [21]:
mylist

[0, 2, 4, 6, 8, 10]

In [22]:
mylist = [x**2 for x in range (0,11) if x%2==0]  

In [23]:
mylist

[0, 4, 16, 36, 64, 100]

In [2]:
celcius = [0,10,24,60,100]

fahrenheit = [((9/5)*temp + 32) for temp in celcius]

In [25]:
fahrenheit

[32.0, 50.0, 75.2, 140.0, 212.0]

In [3]:
fahrenheit = []
for temp in celcius:
    fahrenheit.append (((9/5)*temp + 32))
    

In [4]:
fahrenheit

[32.0, 50.0, 75.2, 140.0, 212.0]

Can also use **if / else** statements inside list comprenhension

In [6]:
results = [ x if x%2==0 else 'ODD' for x in range (0,11)]

In [7]:
results

[0, 'ODD', 2, 'ODD', 4, 'ODD', 6, 'ODD', 8, 'ODD', 10]

In [11]:
success_count = [ 2 if x ==10 else (1 elif x >= 7 else 0) for x in range (0,10)]

SyntaxError: invalid syntax (Temp/ipykernel_12008/2595191980.py, line 1)

Can use **Nested Loops** inside list comprenhension

In [12]:
mylist = []

for x in [2,4,6]:
    for y in [100,200,300]:
        mylist.append (x*y)

In [13]:
mylist

[200, 400, 600, 400, 800, 1200, 600, 1200, 1800]

In [21]:
mylist = []

for x in ['a','b','c']:
    for y in ['d','e','f']:
        mylist.append (x+y)

In [22]:
mylist

['ad', 'ae', 'af', 'bd', 'be', 'bf', 'cd', 'ce', 'cf']

In [38]:
mylist = []
otherlist = ['a','b','c']
for x in otherlist:
    for y in otherlist:
        if x != y:
            mylist.append (x+y)

In [39]:
mylist

['ab', 'ac', 'ba', 'bc', 'ca', 'cb']

In [40]:
mylist = [x*y for x in [2,4,6] for y in [1,10,100]]

In [41]:
mylist

[2, 20, 200, 4, 40, 400, 6, 60, 600]

# METHODS AND FUNCTIONS

### **Methods**

In [67]:
mylist = [1,2,3]

In [68]:
mylist.append(4)

In [69]:
mylist

[1, 2, 3, 4]

In [70]:
mylist.pop()

4

In [71]:
mylist

[1, 2, 3]

To check a bult-in fuction i can **press SHIFT+TAB** after the object.method (only on Jupiter Notebook) **or** run:
**help ( object.method)**


In [77]:
help(mylist.insert)

Help on built-in function insert:

insert(index, object, /) method of builtins.list instance
    Insert object before index.



a **third option** is to check the Python Documentation on python.org or https://docs.python.org/3/

## Functions

.

Using **def** *keyword* to create functions

-> def name_of_function():
        
        '''
        
        Doctring explaning the function
        
        '''
        
        the actual code to be excecuted when the function is called for

In [81]:
def say_hello():
    print("hello")
    print('how are you?')

In [82]:
say_hello()

hello
how are you?


In [83]:
say_hello

<function __main__.say_hello()>

In [88]:
def say_hello(name):
    print(f'Hello {name}')

In [87]:
say_hello('Zalo')

Hello Zalo


In [89]:
say_hello()

TypeError: say_hello() missing 1 required positional argument: 'name'

Here Python asks for the missing parameter. You can go back and complete it, or change the function to have a default value para writing (value='Default_value')

In [94]:
def say_hello(name='nameless stranger'):
    print(f'Hello {name}')

In [95]:
say_hello()

Hello nameless stranger


Use of the **return** *keyword*

In [96]:
def add_num(num1,num2):
    return num1+num2


In [97]:
add_num(10,20)

30

In [98]:
result = add_num(33,65)
print(result)

98


In [99]:
def print_result(a,b):
    print(a+b)

In [100]:
def return_result(a,b):
    return a+b

In [101]:
print_result(10,20)

30


In [102]:
result = print_result(10,20)

30


In [103]:
result

In [104]:
type(result)

NoneType

In [105]:
return_result(10,20)

30

In [106]:
result = return_result(10,20)

In [107]:
result

30

In [108]:
def myfunc(a,b):
    print(a+b)
    return a+b

In [109]:
result = myfunc(10,20)

30


In [110]:
result

30

In [111]:
def sum_numbers(num1,num2):
    return num1+num2

In [112]:
sum_numbers('10','20')

'1020'

In [1]:
2%2

0

In [4]:
3%2

1

In [6]:
20%2==0

True

In [9]:
def even_check(number):
    result = number % 2 == 0
    return result

In [10]:
even_check(20)

True

In [11]:
even_check(21)

False

In [12]:
def even_check(number):
        return number % 2 == 0

In [13]:
even_check(12)

True

In [14]:
# Return True if any number is even inside a list

In [16]:
def check_even_list(num_list):
    for number in num_list:
        if number % 2 == 0:
            return True
        else:
            pass
        

In [19]:
check_even_list([1,3,5,7])

In [20]:
check_even_list([1,3,6,7])

True

In [21]:
def check_even_list(num_list):
    for number in num_list:
        if number % 2 == 0:
            return True
        else:
            return False # WRONG!!! Because it will stop at the first odd number and not check the rest
        

In [22]:
check_even_list([1,3,6,7])

False

In [23]:
def check_even_list(num_list):
    for number in num_list:
        if number % 2 == 0:
            return True
        else:
            pass
    return False #It has to be in the same indentation that the FOR loop

In [24]:
check_even_list([1,3,6,7])

True

In [25]:
check_even_list([1,3,5,7])

False

In [26]:
def check_even_list(num_list):
    # return all the even number in a list
    
    #placeholder variables:
    even_numbers = []
        
    
    for number in num_list:
        if number % 2 == 0:
            even_numbers.append(number)
        else:
            pass
    return even_numbers

In [27]:
check_even_list([1,2,3,4,5,6,7,8,9,10])

[2, 4, 6, 8, 10]

In [28]:
check_even_list([1,3,5,7])

[]


**Tuples in def funct**


In [29]:
stock_prices = [('APPL', 200), ('GOOG',400),('MSFT',100)]

In [30]:
for item in stock_prices:
    print (item)

('APPL', 200)
('GOOG', 400)
('MSFT', 100)


In [32]:
for ticker,price in stock_prices:
    print(price+(0.1*price))

220.0
440.0
110.0


In [45]:
work_hours = [('Abby',100),('Billy',400),('Cassie',800),('Danny', 600)]

In [46]:
def employee_check (work_hours):
    #placeholders
    current_max=0
    employee_of_month = ''
    
    for employee,hours in work_hours:
        if hours > current_max:
            current_max = hours
            employee_of_month = employee
        else:
            pass
    
    # Return
    return (employee_of_month, current_max)

In [47]:
result = employee_check(work_hours)

In [48]:
result

('Cassie', 800)

In [49]:
name,hours = employee_check(work_hours)

In [50]:
result

('Cassie', 800)

In [51]:
name

'Cassie'

In [52]:
hours

800